kubernetes_asyncio/leaderelection/leaderelection.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import datetime
+import json
+import logging
+import sys
+import time
+from http import HTTPStatus
+
+from .leaderelectionrecord import LeaderElectionRecord
+
+logging.basicConfig(level=logging.INFO)
+
+"""
+This package implements leader election using an annotation in a Kubernetes
+object. The onstarted_leading coroutine is run as a task, which is cancelled if
+the leader lock is obtained and then lost.
+
+At first all candidates are considered followers. The one to create a lock or
+update an existing lock first becomes the leader and remains so until it keeps
+renewing its lease.
+"""
+
+
+class LeaderElection:
+    def __init__(self, election_config):
+        if election_config is None:
+            sys.exit("argument config not passed")
+
+        # Latest record observed in the created lock object
+        self.observed_record = None
+
+        # The configuration set for this candidate
+        self.election_config = election_config
+
+        # Latest update time of the lock
+        self.observed_time_milliseconds = 0
+
+    # Point of entry to Leader election
+    async def run(self):
+        # Try to create/ acquire a lock
+        if await self.acquire():
+            logging.info(
+                "{} successfully acquired lease".format(
+                    self.election_config.lock.identity
+                )
+            )
+
+            task = asyncio.create_task(self.election_config.onstarted_leading())
+
+            await self.renew_loop()
+
+            # Leader lock lost - cancel the onstarted_leading coroutine if it's
+            # still running
+            task.cancel()
+
+            # Failed to update lease, run OnStoppedLeading callback
+            await self.election_config.onstopped_leading()
```

Comment:
The implementation from `kubernetes-client/python` uses threads without a cancellation equivalent, so `onstopped_leading` is provided as an optional completion callback. In this implementation `onstarted_leading` is cancelled if the leader lock is lost, to allow that function to clean up state that might not be accessible to a `onstopped_leading` callback coroutine.

Do you think there's still a purpose for `onstopped_leading` (including maintaining a similar interface to the other library)? Or should it be removed in favor of handling cancellation?

Reply:
I think it'd be better to keep this to have a similar interface.

Reply:
I kept the `onstopped_leading`, but also preserved the `onstarted_leading` cancellation.